File: tlg/src/ch/bailu/tlg/MatrixVisible.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;

public class MatrixVisible extends Matrix {

    private TlgRectangle dirtyArea = new TlgRectangle(0,0,0,0);
    private boolean drawGrid=false;

    public MatrixVisible(int w, int h) {
        super(w, h);
        markAllAsDirty();
    }

    public MatrixVisible(Matrix m) {
        super(m);
        markAllAsDirty();
    }
    
    public MatrixVisible(BufferedInputStream reader) throws IOException {
        super(reader);
        if (reader.read() > 0) drawGrid=true;
    }
    
    public void writeState(BufferedOutputStream writer) throws IOException {
        super.writeState(writer);
        if (drawGrid) writer.write(1);
        else writer.write(0);
    }

    public void erase() {
        for (int i=0; i<count(); i++) {
            get(i).makeInvisible();
        }
        markAllAsDirty();
    }
    
    protected void markAllAsDirty() {
        dirtyArea.left=0;
        dirtyArea.top=0;
        dirtyArea.setWidth(getWidth());
        dirtyArea.setHeight(getHeight());
    }

    protected void markAllAsClean() {
        dirtyArea.left=getWidth()-1;
        dirtyArea.top=getHeight()-1;
        dirtyArea.bottom=0; //setWidth(0);
        dirtyArea.right=0; //setHeight(0);
    }
    
    
    protected Square getD(int x, int y) {
        markSquareAsDirty(x,y);
        return get(x,y);
    }
    
    private void markSquareAsDirty(int x, int y) {
        if (x < dirtyArea.left)  dirtyArea.left = x;
        if (y < dirtyArea.top)  dirtyArea.top = y;
     
        if (x > dirtyArea.right)  dirtyArea.right  = x;
        if (y > dirtyArea.bottom) dirtyArea.bottom = y;
    }
    
    void  setPixelGeometry(TlgRectangle rect) {
        int   xoffset,yoffset,xpos,ypos;

        int squareSize = rect.getWidth() / getWidth();
        if (squareSize > rect.getHeight() / getHeight()) squareSize = rect.getHeight() / getHeight();

        // center
        xoffset = rect.left + (rect.getWidth()  - (squareSize * getWidth()) ) / 2;  
        yoffset = rect.top + (rect.getHeight() - (squareSize * getHeight()) ) / 2;  

        xpos=xoffset;
        for (int x = 0; x < getWidth(); x++) {

            ypos=yoffset;
            for (int y = 0; y < getHeight(); y++) {
                
                get(x,y).rect = new TlgRectangle(xpos, ypos, xpos+squareSize-1, ypos+squareSize-1);
                ypos+=squareSize;
            }
            xpos+=squareSize;
        }
        markAllAsDirty();
    }
    

    public void toggleGrid() {
        drawGrid = !drawGrid;
        markAllAsDirty();
    }
    
    public TlgRectangle getPixelGeometry() {
        return getPixelArea(new TlgRectangle(0,0,getWidth()-1,getHeight()-1));
    }
    
    private TlgRectangle getDirtyPixelArea() {
        return getPixelArea(dirtyArea);
    }
    
    private TlgRectangle getPixelArea(TlgRectangle rect) {
        TlgRectangle ret= new TlgRectangle(0,0,0,0);
        
        if (rect.getWidth() > 0 && rect.getHeight() > 0) {
            Square tl=get(rect.left,rect.top);
            Square br=get(rect.right, rect.bottom);
            
            ret.left=tl.rect.left;
            ret.top=tl.rect.top;
            ret.right=br.rect.right;
            ret.bottom=br.rect.bottom;
        }
        return ret;
    }
    
    private boolean isDirty() {
        return dirtyArea.getWidth() > 0 && dirtyArea.getHeight() > 0;
    }
    
    public void update(PlatformContext gc) {
        if (isDirty()) {
            gc.setDirtyRect(getDirtyPixelArea());
        
            for (int x=dirtyArea.left; x <= dirtyArea.right; x++) { 
                for (int y=dirtyArea.top; y <= dirtyArea.bottom; y++) {
                    get(x,y).update(gc, drawGrid);
                }
            }
        }
        markAllAsClean();
    }
    
    public void updateAll(PlatformContext gc) {
        TlgRectangle r=getPixelGeometry();
        r.grow(1);
        gc.drawRectangle(gc.colorFrame(), r);
        
        markAllAsDirty();
        update(gc);
    }
}

File: tlg/src/main/kotlin/tlg/matrix/MatrixVisible.kt
package tlg.matrix

import tlg.Square
import tlg.context.PlatformContext
import tlg.geometry.TlgRectangle
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException

open class MatrixVisible : Matrix {
    private val dirtyArea: TlgRectangle = TlgRectangle()
    private var drawGrid = false

    constructor(w: Int, h: Int) : super(w, h) {
        markAllAsDirty()
    }

    constructor(reader: BufferedInputStream) : super(reader) {
        if (reader.read() > 0) drawGrid = true
    }

    @Throws(IOException::class)
    override fun writeState(out: BufferedOutputStream) {
        super.writeState(out)
        if (drawGrid) out.write(1) else out.write(0)
    }

    fun erase() {
        for (i in 0 until count()) {
            get(i).makeInvisible()
        }
        markAllAsDirty()
    }

    protected fun markAllAsDirty() {
        dirtyArea.left = 0
        dirtyArea.top = 0
        dirtyArea.width = width
        dirtyArea.height = height
    }

    protected fun markAllAsClean() {
        dirtyArea.left = width - 1
        dirtyArea.top = height - 1
        dirtyArea.bottom = 0 //setWidth(0);
        dirtyArea.right = 0 //setHeight(0);
    }

    fun getD(x: Int, y: Int): Square {
        markSquareAsDirty(x, y)
        return get(x, y)
    }

    private fun markSquareAsDirty(x: Int, y: Int) {
        if (x < dirtyArea.left) dirtyArea.left = x
        if (y < dirtyArea.top) dirtyArea.top = y
        if (x > dirtyArea.right) dirtyArea.right = x
        if (y > dirtyArea.bottom) dirtyArea.bottom = y
    }

    fun toggleGrid() {
        drawGrid = !drawGrid
        markAllAsDirty()
    }

    // center
    var pixelGeometry: TlgRectangle
        get() = getPixelArea(TlgRectangle(0, 0, width - 1, height - 1))
        set(rect) {
            var squareSize: Int = rect.width / width
            if (squareSize > rect.height / height) {
                squareSize = rect.height / height
            }

            // center
            val xOffset = rect.left + (rect.width - squareSize * width) / 2
            val yOffset = rect.top + (rect.height - squareSize * height) / 2
            var xpos: Int = xOffset
            var ypos: Int
            for (x in 0 until width) {
                ypos = yOffset
                for (y in 0 until height) {
                    get(x, y).rect =
                        TlgRectangle(xpos, ypos, xpos + squareSize - 1, ypos + squareSize - 1)
                    ypos += squareSize
                }
                xpos += squareSize
            }
            markAllAsDirty()
        }

    private val dirtyPixelArea: TlgRectangle
        get() = getPixelArea(dirtyArea)

    private fun getPixelArea(rect: TlgRectangle): TlgRectangle {
        val result = TlgRectangle(0, 0, 0, 0)
        if (rect.width > 0 && rect.height > 0) {
            val tl = get(rect.left, rect.top)
            val br = get(rect.right, rect.bottom)
            result.left = tl.rect.left
            result.top = tl.rect.top
            result.right = br.rect.right
            result.bottom = br.rect.bottom
        }
        return result
    }

    private val isDirty: Boolean
        get() = dirtyArea.width > 0 && dirtyArea.height > 0

    fun update(pContext: PlatformContext) {
        if (isDirty) {
            pContext.setDirtyRect(dirtyPixelArea)
            for (x in dirtyArea.left..dirtyArea.right) {
                for (y in dirtyArea.top..dirtyArea.bottom) {
                    get(x, y).update(pContext, drawGrid)
                }
            }
        }
        markAllAsClean()
    }

    fun updateAll(pContext: PlatformContext) {
        val r: TlgRectangle = pixelGeometry
        r.grow(1)
        pContext.drawRectangle(pContext.colorFrame(), r)
        markAllAsDirty()
        update(pContext)
    }
}
